### PR TITLE
fix(goal-funding): exclude future-dated deposits from computeMonthlyDepositRate

### LINKED ISSUE #257: tech debt: computeMonthlyDepositRate includes future-dated deposits, inflating projected rate
`src/lib/goal-funding.ts` — `computeMonthlyDepositRate` filters transactions by type (`Deposit`) but does not filter by date. If the ledger contains deposits with a date after `referenceDate` (possible via the deposit dialog, which accepts any date), those deposits are counted in the total and in the earliest-deposit window calculation. This can inflate the computed rate and produce ETAs that are earlier than realistic.

**Expected behavior**: Only deposits where `tx.date <= referenceDate` should be included in the rate calculation. Both the sum and the earliest-deposit lookup should operate on past deposits only.

**Actual behavior**: All deposits (including future-dated) are included.

**Fix**: Add a `tx.date <= referenceDate` filter before the `Deposit` type filter in `computeMonthlyDepositRate`.

The deposit dialog (`src/components/ledgers/AddDepositDialog.tsx`) uses an unrestricted `<input type="date">` with no `max` attribute, so future-dated deposits are possible in practice.

Identified during review of [#246](https://github.com/rmartz/personal-budget/pull/246).

---
*Created by Claude Sonnet 4.6*

### PR DESCRIPTION
**Pull request by rmartz**
The deposit dialog accepts any date, so users can create deposits dated in the future. Previously, `computeMonthlyDepositRate` included all deposits regardless of date, which inflated the monthly average and produced unrealistically optimistic goal ETAs.

The fix adds a `tx.date <= referenceDate` guard to the deposit filter, excluding future-dated entries from both the deposit total and the earliest-deposit anchor used to compute the time window. This ensures only realized, historical deposits inform the projected rate.

Closes #257

---
*Created by Claude Sonnet 4.5*

### CODE REVIEW

**Review by rmartz — COMMENTED** [review]
## Requested changes

- **[src/lib/goal-funding.spec.ts line 145](https://github.com/rmartz/personal-budget/pull/263#discussion_r3291658703)**: The test "does not use a future deposit as the earliest-deposit anchor" has an internally inconsistent comment. The inline note says "its earlier calendar month (Jul) would expand the window," but the future deposit in the test is dated January 2026, not July — that reference appears to be a leftover from an earlier draft. The second half of the comment then acknowledges the scenario isn't accurately represented ("a future deposit after REF would not be 'earlier' than Jan anyway"), meaning the test name overstates what the test validates. Both the test name and comment should be revised to accurately describe what is being asserted: that a future-dated deposit is excluded from both the sum and the earliest-deposit anchor, not the specific anchor-expansion scenario the name implies.

## Verdict

One test comment has a factual inconsistency ("Jul" when using January 2026) and the test name overpromises relative to the scenario actually constructed. The implementation is correct and all acceptance criteria from [#257](https://github.com/rmartz/personal-budget/issues/257) are met; only the test documentation needs to be corrected before merge.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Review

**CI**: All checks pass. The third commit correctly addresses the prior review's request to fix the inconsistent test name and comment.

**Implementation correctness**: The core logic — filtering `tx.date <= referenceDate` before computing the total and the earliest-deposit anchor — is correct for the stated goal. The window computation and the docstring update are both good.

**Outstanding issue**: Copilot's second-round comments (on the latest commit) raise a real correctness concern that still needs to be addressed before merge.

### Timezone mismatch between deposit dates and `referenceDate`

Tracing the data path:

1. `AddDepositDialog` collects a date string (e.g. `"2025-06-16"`) and calls `new Date(date)` → parses as **UTC midnight** (`2025-06-16T00:00:00.000Z`).
2. `budgetLedgerTransactionToFirebase` stores it via `.toISOString()` and `firebaseToBudgetLedgerTransaction` restores it with `new Date(data.date)` — both sides remain UTC midnight.
3. `computeMonthlyDepositRate` defaults `referenceDate` to `new Date()` — a **local-time timestamp** with a full time component.

In a **negative-UTC-offset timezone** (e.g. US/Pacific = UTC-7), when the local clock reads June 15 at 23:00, `new Date()` is `2025-06-16T06:00:00Z`. A deposit dated June 16 is stored as `2025-06-16T00:00:00Z`. The guard evaluates `2025-06-16T00:00:00Z <= 2025-06-16T06:00:00Z` → **true** — the future-dated deposit slips through.

The fix is to compare by calendar day rather than raw timestamp: normalize `referenceDate` to end-of-day UTC (or compare using UTC year/month/day) before filtering, or truncate `referenceDate` to start-of-day local time. The month-based window computation below already uses `getFullYear()`/`getMonth()` (calendar-safe), so aligning the filter to the same calendar-day boundary is consistent.

The test coverage gap (no test that mirrors `new Date("YYYY-MM-DD")` deposit dates against a `referenceDate` with a time component) means this class of bug won't be caught by the current test suite — that test addition is needed alongside the fix.

Not approving until this is addressed.

---
*Posted by Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Review

**CI**: All checks pass.

**Timezone fix**: The new commit correctly resolves the timezone mismatch flagged in the previous review. The approach is sound:

- `refDayUTC` is built with `Date.UTC(referenceDate.getFullYear(), referenceDate.getMonth(), referenceDate.getDate())` — reads the **local calendar day** of `referenceDate` and converts it to a UTC midnight timestamp. This is the right choice because `new Date()` represents local time and users think in local calendar days.
- Deposit dates are compared using `Date.UTC(tx.date.getUTCFullYear(), tx.date.getUTCMonth(), tx.date.getUTCDate())` — reads the **UTC fields** because deposit dates are stored as UTC midnight ISO strings. This is also correct.

The asymmetry (local getters for `referenceDate`, UTC getters for deposit dates) is intentional and necessary — it is the exact normalization needed to align two dates that live in different time contexts onto the same calendar-day coordinate system.

**Test coverage**: The two new tests cover the key scenarios — a same-day UTC-midnight deposit included when `referenceDate` has a time component, and a next-day UTC-midnight deposit excluded. These directly address the gap Copilot flagged.

**Verdict**: All previously blocking issues are resolved. The implementation is correct, CI is green, and the test suite covers the timezone edge cases. Ready to merge.

---
*Posted by Claude Sonnet 4.6*

File: src/lib/goal-funding.ts [hunk 1]
```diff
@@ -11,13 +11,33 @@ import type { BudgetLedgerSavingsGoal } from "@/lib/firebase/schema/savings-goal
...
  */
 export function computeMonthlyDepositRate(
   transactions: BudgetLedgerTransaction[],
   referenceDate: Date = new Date(),
 ): number {
+  const refDayUTC = Date.UTC(
+    referenceDate.getFullYear(),
+    referenceDate.getMonth(),
+    referenceDate.getDate(),
+  );
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/lib/goal-funding.ts:38]: Pre-existing code outside this PR's scope — the `getFullYear()/getMonth()` window calculation was not introduced by this PR; only the deposit filter was changed. The asymmetry between UTC-based filtering and local-time window counting is tracked as tech debt in #270. — *Claude Sonnet 4.6*

File: src/lib/goal-funding.spec.ts [hunk 2]
```diff
@@ -104,6 +104,88 @@ describe("computeMonthlyDepositRate", () => {
       expect(computeMonthlyDepositRate(transactions, REF_DATE)).toBe(240);
     });
   });
+
+  describe("excludes future-dated deposits", () => {
+    it("returns 0 when all deposits are after the reference date", () => {
+      // Only future deposits (Jul 2025) — nothing to base a rate on
+      const transactions = [
+        makeTransaction({ amount: 600, date: new Date(2025, 6, 1) }),
+      ];
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/lib/goal-funding.spec.ts:118]: Pre-existing code concern — the tests added by this PR correctly verify filter behavior; the inconsistency between local-midnight test dates and production UTC-midnight dates is a symptom of the pre-existing window calculation issue, now tracked in #270. Fixing the test dates requires fixing the window calculation first (otherwise the new expected values would themselves be timezone-dependent). — *Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ Timezone mismatch between deposit dates and `referenceDate` — fixed in `cfa6bec`. The `refDayUTC` approach (local-calendar-day getters on `referenceDate`, UTC getters on deposit dates) is correct and intentional. Two new tests cover the key UTC-midnight scenarios.

## Verdict

All prior items resolved. The two open Copilot threads ([goal-funding.ts line 32](https://github.com/rmartz/personal-budget/pull/263#discussion_r3294838891) and [goal-funding.spec.ts line 113](https://github.com/rmartz/personal-budget/pull/263#discussion_r3294838895)) both surface the same pre-existing issue — the `startMonths` window calculation uses local-time getters (`getFullYear()/getMonth()`) for UTC-midnight deposit dates — which is code this PR did not change. Both threads have been dismissed with explanation. The two new commits since the last review are a branch-update merge (pulling main's changes from PR [#244](https://github.com/rmartz/personal-budget/pull/244)); no changes to `goal-funding.ts` or `goal-funding.spec.ts`. Ready to merge.

## Deferred

- **Window calculation timezone inconsistency** — `startMonths` derives from `earliestDeposit.date.getFullYear()/getMonth()` (local time), while the filter introduced by this PR uses UTC getters. For UTC-midnight deposit dates in negative UTC-offset timezones, local getters can return the previous calendar day, inflating `monthsElapsed` by one month and yielding a lower-than-actual monthly rate. Pre-existing code; the deposit-rate tests also use local-midnight dates rather than UTC-midnight, masking the issue. Tracked in [#270](https://github.com/rmartz/personal-budget/issues/270).

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
The two commits since the last review are branch-update merges (pulling main's changes from PR 262). The incremental diff introduces the cashCap optional parameter to computeMonthlyDepositRate, a per-deposit ceiling applied before summing, plus three tests covering above-cap clamping, below-cap pass-through, and parity with the no-cap call. Implementation is correct and backward-compatible (undefined default preserves existing behavior). No issues found. Issue 257 acceptance criteria remain fully met. CI is green. All prior items resolved and no new issues - ready to merge.

---

_Claude Sonnet 4.6_

**Comment by rmartz** [comment]
**Pre-merge check**

- ✅ No WIP or draft
- ✅ PR approved, no changes requested, no open threads, reviewed by /review
- ✅ UAT waived (skip-uat)
- ✅ Targets default branch, up to date, no conflicts
- ✅ All CI green

Merging.

---

_Claude Sonnet 4.6_
